File: python/supersrg/srg.py
from io import TextIOBase
from . import Mappings

class SrgMappingsEncoder:
    __slots__ = "output"
    def __init__(self, output: TextIOBase):
        self.output = output

    def encode(self, mappings: Mappings):
        output = self.output
        for original, renamed in mappings.classes.items():
            output.write("CL: ")
            output.write(original.internal_name)
            output.write(" ")
            output.write(renamed.internal_name)
            output.write("\n")
        for original, renamed in mappings.fields.items():
            output.write("FD: ")
            output.write(original.declaring_class.internal_name)
            output.write('/')
            output.write(original.name)
            output.write(" ")
            output.write(renamed.declaring_class.internal_name)
            output.write('/')
            output.write(renamed.name)
            output.write("\n")
        for original, renamed in mappings.methods.items():
            output.write("MD: ")
            output.write(original.declaring_class.internal_name)
            output.write('/')
            output.write(original.name)
            output.write(' ')
            output.write(original.signature.descriptor)
            output.write(' ')
            output.write(renamed.declaring_class.internal_name)
            output.write('/')
            output.write(renamed.name)
            output.write(' ')
            output.write(renamed.signature.descriptor)
            output.write('\n')
    

File: native/src/mappings/encoder/srg.rs
use std::io::{self, Write};

use super::MappingsEncoder;
use mappings::{MappingsSnapshot, MappingsIterator};
use types::{JavaClassLookup, FieldDataLookup, MethodDataLookup};

pub struct SrgEncoder<'a> {
    mappings: &'a MappingsSnapshot,
}
impl<'a> MappingsEncoder<'a> for SrgEncoder<'a> {
    #[inline]
    fn new(mappings: &'a MappingsSnapshot) -> Self {
        SrgEncoder { mappings }
    }
    #[inline]
    fn write<W: Write>(&self, out: &mut W) -> io::Result<()> {
        for (original, renamed) in self.mappings.classes() {
            out.write_all(b"CL: ")?;
            out.write_all(original.internal_name().as_bytes())?;
            out.write_all(b" ")?;
            out.write_all(renamed.internal_name().as_bytes())?;
            out.write_all(b"\n")?;
        }
        for (original, renamed) in self.mappings.fields() {
            out.write_all(b"FD: ")?;
            out.write_all(original.class().internal_name().as_bytes())?;
            out.write_all(b"/")?;
            out.write_all(original.name().as_bytes())?;
            out.write_all(b" ")?;
            out.write_all(renamed.class().internal_name().as_bytes())?;
            out.write_all(b"/")?;
            out.write_all(renamed.name().as_bytes())?;
            out.write_all(b"\n")?;
        }
        for (original, renamed) in self.mappings.methods() {
            out.write_all(b"MD: ")?;
            out.write_all(original.class().internal_name().as_bytes())?;
            out.write_all(b"/")?;
            out.write_all(original.name().as_bytes())?;
            out.write_all(b" ")?;
            out.write_all(original.signature().as_bytes())?;
            out.write_all(b" ")?;
            out.write_all(renamed.class().internal_name().as_bytes())?;
            out.write_all(b"/")?;
            out.write_all(renamed.name().as_bytes())?;
            out.write_all(b" ")?;
            out.write_all(renamed.signature().as_bytes())?;
            out.write_all(b"\n")?;
        }
        Ok(())
    }
}
